src/app/facebook-pixel-service/facebook-pixel-service.ts

```diff
@@ -0,0 +1,49 @@
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class FacebookPixelService {
+  private loadOk = false;
+
+  constructor() {}
+
+  load() {
+    if (!this.loadOk) {
+      ((f: any, b, e, v, n, t, s) => {
+        if (f.fbq) {
+          return;
+        }
+        n = f.fbq = () => {
```

Comment:
@NoelLH Any ideas on this error here: `TS2322: Type '() => void' is not assignable to type 'undefined'` - also getting `TS2532: Object is possibly 'undefined'` for every param variable assignment

Reply:
@tbg-jasonfung if you look at where the error is first highlighted it's the assignment to `n` which the FB snippet uses extensively. So you need to tell TypeScript a compatible type for `n`.

FB's JS is not in our control and we don't know or care what types it uses (and they are liable to change over time), so the only appropriate type is `any`.

Put that in and you'll find there is a new error because the anon fn they tell you to declare takes up to 7 args but we only pass in 4 on the initial call. So we need to tell TypeScript it's an _optional_ param of type `any`: `n?: any`

Repeat for the other params.

Reply:
@NoelLH Thank you, that's really helpful - its now compiling okay